### PR TITLE
Added Process ID to sloggers as optional KV 

### PR DESCRIPTION
**Pull request by przygienda**

### CODE REVIEW

File: src/file.rs [hunk 1]
```diff
@@ -25,6 +25,7 @@ use {Build, Config, ErrorKind, Result};
 pub struct FileLoggerBuilder {
     format: Format,
     source_location: SourceLocation,
+    processid: ProcessID,
```
Comment (sile) [inline src/file.rs]: I think `process_id` and `ProcessId` are more suited with [the Rust naming conventions](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html).

File: src/file.rs [hunk 2]
```diff
@@ -152,20 +160,40 @@ impl FileLoggerBuilder {
 
             let drain = self.level.set_level_filter(kvdrain.fuse());
 
-            match self.source_location {
-                SourceLocation::None => Logger::root(drain.fuse(), o!()),
-                SourceLocation::ModuleAndLine => {
-                    Logger::root(drain.fuse(), o!("module" => FnValue(module_and_line)))
-                }
+            match (self.source_location, self.processid) {
```
Comment (sile) [inline src/file.rs]: Because this code, [file.rs#L182](https://github.com/sile/sloggers/pull/20/files#diff-902ed5e7647ce8bdec1b24ca4254e1faR182), [terminal.rs#L113](https://github.com/sile/sloggers/pull/20/files#diff-6d00071c5cb83f348e478f2a87d45999R113) and [terminal.rs#L132](https://github.com/sile/sloggers/pull/20/files#diff-6d00071c5cb83f348e478f2a87d45999R132) seem too redundant, it may be better to create a function to factor out this code and reuse it.

File: src/file.rs [hunk 3]
```diff
@@ -406,6 +434,10 @@ pub struct FileLoggerConfig {
     #[serde(default)]
     pub source_location: SourceLocation,
 
+    /// process ID
```
Comment (sile) [inline src/file.rs:440]: [Rust Style Guidelines](https://doc.rust-lang.org/1.0.0/style/style/comments.html#sentence-structure) says,
> All doc comments, including the summary line, should begin with a capital letter and end with a period, question mark, or exclamation point.

File: src/terminal.rs [hunk 4]
```diff
@@ -50,6 +52,12 @@ impl TerminalLoggerBuilder {
         self
     }
 
+    /// Sets the source code location type this logger will use.
```
Comment (sile) [inline src/terminal.rs]: This description seems inappropriate (copy and paste mistake?).

File: src/types.rs [hunk 5]
```diff
@@ -220,3 +220,41 @@ impl FromStr for SourceLocation {
...
+    fn default() -> Self {
+        ProcessID::ProcessID
+    }
+}
+impl FromStr for ProcessID {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<Self, Error> {
+        match s {
+            "none" => Ok(ProcessID::None),
+            "id" => Ok(ProcessID::ProcessID),
```
Comment (sile) [inline src/types.rs]: This key should match with the representation of `serde` (i.e., "process_id").

File: src/types.rs [hunk 6]
```diff
@@ -220,3 +220,41 @@ impl FromStr for SourceLocation {
...
+#[allow(missing_docs)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum ProcessID {
+    None,
+    ProcessID,
+}
+impl Default for ProcessID {
+    fn default() -> Self {
+        ProcessID::ProcessID
```
Comment (sile) [inline src/types.rs]: I think the default value should be `ProcessID::None` to keep compatibility with past versions.

File: src/terminal.rs [hunk 7]
```diff
@@ -221,6 +249,10 @@ pub struct TerminalLoggerConfig {
     #[serde(default)]
     pub source_location: SourceLocation,
 
+    /// process ID
+    #[serde(default)]
+    pub processid: ProcessID,
```
Comment (sile) [inline src/terminal.rs]: I think , in this setting, the associated configuration files will be slightly clunky:
For example,
```toml
type = "terminal" 
source_location = "module_and_line" 
level = "debug"

# Inlcude PID in log messages
processid = "process_id"
```

Hence, it may be better to replace the value type of `processid` to `bool`.
This will allow you to write as the following:
```toml
type = "terminal" 
source_location = "module_and_line" 
level = "debug"

# Inlcude PID in log messages
processid = true
```

**Review by sile — CHANGES_REQUESTED** [review]
Thank you for your PR. It seems a good feature. 
I wrote some comments, so please check them out.
Besides I updated `master` branch to fix CI failures. Please merge it.

File: src/file.rs (hunk 1 above)
Comment (przygienda) [reply src/file.rs]: will do

File: src/file.rs (hunk 2 above)
Comment (przygienda) [reply src/file.rs]: sigh, yes, all those branches are generating a different 

o!() variant and hence every branch much fuse() the logger doing unnecessary code repetition ... I'll see whether I can move that out into some kind of fn()

File: src/file.rs (hunk 3 above)
Comment (przygienda) [reply src/file.rs:440]: will do

File: src/terminal.rs (hunk 4 above)
Comment (przygienda) [reply src/terminal.rs]: will do

File: src/types.rs (hunk 5 above)
Comment (przygienda) [reply src/types.rs]: will do

File: src/types.rs (hunk 6 above)
Comment (przygienda) [reply src/types.rs]: fair enough

File: src/terminal.rs (hunk 7 above)
Comment (przygienda) [reply src/terminal.rs]: done

File: src/file.rs [hunk 8]
```diff
@@ -33,68 +33,75 @@ pub struct FileLoggerBuilder {
 }
 
 impl FileLoggerBuilder {
-    /// Makes a new `FileLoggerBuilder` instance.
+	/// Makes a new `FileLoggerBuilder` instance.
```
Comment (sile) [inline src/file.rs]: It seems that the indentations of this file are broken. Could you re-format this file by using `rustfmt` (i.e., `$ cargo fmt`)?

File: src/misc.rs [hunk 9]
```diff
@@ -18,9 +22,66 @@ pub fn module_and_line(record: &Record) -> String {
...
+}
+
 pub fn timezone_to_timestamp_fn(timezone: TimeZone) -> fn(&mut io::Write) -> io::Result<()> {
     match timezone {
         TimeZone::Utc => slog_term::timestamp_utc,
         TimeZone::Local => slog_term::timestamp_local,
     }
 }
+
+/// builds a logger based on given combination options
```
Comment (sile) [inline src/misc.rs:36]: Please see https://github.com/sile/sloggers/pull/20/files#r245470559.

File: src/types.rs [hunk 10]
```diff
@@ -220,3 +220,40 @@ impl FromStr for SourceLocation {
...
+///
+/// ```
+/// use sloggers::types::ProcessID;
+///
+/// assert_eq!(ProcessID::default(), ProcessID(false));
+/// ```
+#[allow(missing_docs)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct ProcessID(pub bool);
```
Comment (sile) [inline src/types.rs:238]: In my opinion, it is excessive to define a new struct for this purpose (simple boolean type is sufficient).

Comment (przygienda) [reply src/types.rs:238]: I disagree, encapsulation of types prevents weird problems like multiple parameters of same type being confused in sequence and so on. Plus, it looks more consistsent with the other features

File: src/file.rs (hunk 8 above)
Comment (przygienda) [reply src/file.rs]: done

File: src/misc.rs (hunk 9 above)
Comment (przygienda) [reply src/misc.rs:36]: hmm, formatting again I assume, ran through rustfmt

**Comment by przygienda** [comment]
I also added file syn'ing on drop. The logger to file looses regularly entries now on shutdown.

File: src/types.rs (hunk 10 above)
Comment (sile) [reply src/types.rs:238]: > encapsulation of types prevents weird problems

Generally, I agree with your opinion. However, wrapping every basic types  feels to be cumbersome (e.g., users are required to import many types to use the library). Besides, in this case of the process ID option, I think a subtle situation such as "multiple parameters of same type being confused in sequence" will hardly occur. 

> it looks more consistsent with the other features

Currently, `sloggers` uses bare types for simple options (e.g., [`rotate_compress`](https://docs.rs/sloggers/0.3.1/sloggers/file/struct.FileLoggerConfig.html#structfield.rotate_compress), [`truncate`](https://docs.rs/sloggers/0.3.1/sloggers/file/struct.FileLoggerConfig.html#structfield.truncate), [`channel_size`](https://docs.rs/sloggers/0.3.1/sloggers/file/struct.FileLoggerConfig.html#structfield.channel_size)).

File: src/file.rs [hunk 11]
```diff
@@ -189,10 +180,49 @@ impl Build for FileLoggerBuilder {
     }
 }
 
+/// Files on drop do not sync which cuts of tail-end of the logs. Encaps the type to make
```
Comment (sile) [inline src/file.rs:183]: Although  this seems useful fixing, it's unrelated to the subject of this PR. I would appreciate if you could split this modification to another PR.

Comment (przygienda) [reply src/file.rs:183]: will do in a bit, need that code right now on my branch working ...

**Comment by sile** [comment]
Since there are no activities on this PR for a long time, please let me close it. If you have any progress, feel free to reopen it or create another PR.
